File: pkg/docstate/adopt.go
package docstate

import (
	"database/sql"
	"fmt"
	"time"
)

// ResolveAdopt commits a [D]iscard/[M]erge resolution (or an explicit
// hash-equality adopt): it re-tags the observation obs (already recorded —
// theirs was captured by I1 the moment it was read) as origin='resolve',
// correlated to editSeq (the seq of the journaled ReplaceAll/merge-entry edit
// that resolved it), and advances saved_obs to it — all in ONE short tx, pure
// SQLite. Undo past editSeq moves the journal position below this resolve
// observation, so ancestorAt automatically stops finding it and Sync reports
// Diverged again — the guard re-raises without any bespoke unwind logic. The
// new row's supersedes records whatever saved_obs held immediately before
// this call, so a later ResolveAbandon (an Esc-abort of the resolution) can
// restore it EXACTLY.
func (s *Store) ResolveAdopt(docID int64, obs ObsID, editSeq int64) error {
	source, err := s.getObservation(obs)
	if err != nil {
		return fmt.Errorf("resolve adopt doc %d: read source observation: %w", docID, err)
	}
	// Routed through recordAdoption (the shared one-tx primitive) instead of
	// reimplementing "read prior saved_obs, insert, advance saved_obs" inline.
	if _, err := s.recordAdoption(docID, source.BlobHash, source.Size, source.MTime, source.Inode, source.Device, source.NLink, "resolve", editSeq); err != nil {
		return fmt.Errorf("resolve adopt doc %d: %w", docID, err)
	}
	return nil
}

// ResolveAbandon reverses the resolve observation a merge/discard adoption
// (ResolveAdopt) or a heal-adopt created — the store half of an Esc-abort out
// of the merge resolver (F3). It reads docID's CURRENT saved_obs (expected to
// be the resolve/adoption observation the caller is unwinding), deletes that
// row (the blob itself is kept — history is never destroyed, only the fact
// that it was "agreed" is retracted), and restores saved_obs to EXACTLY what
// it superseded — never re-derived by an origin scan, which an intervening
// diverged 'load'/'probe' sighting recorded between the adoption and this
// abandon could poison into picking the wrong baseline (critic R4). With
// theirs now derived from the newest sighting (Sync's Adoption Contract),
// whatever external bytes this resolution reconciled are still visible after
// the deletion — Sync immediately re-reports Diverged (or whatever it
// reported before the resolution) with no separate re-derivation needed here.
// A docID with no saved_obs at all (never adopted anything) is a safe no-op.
func (s *Store) ResolveAbandon(docID int64) error {
	tx, err := s.perm.Begin()
	if err != nil {
		return fmt.Errorf("resolve abandon doc %d: begin: %w", docID, err)
	}
	var current sql.NullInt64
	err = tx.QueryRow(`SELECT saved_obs FROM session_documents WHERE session_id=? AND doc_id=?`, s.sessionID, docID).Scan(&current)
	if err == sql.ErrNoRows {
		tx.Rollback() //nolint:errcheck
		return nil    // no session_documents row at all — nothing adopted yet, safe no-op
	}
	if err != nil {
		tx.Rollback() //nolint:errcheck
		return fmt.Errorf("resolve abandon doc %d: read saved_obs: %w", docID, err)
	}
	if !current.Valid {
		tx.Rollback() //nolint:errcheck
		return nil    // nothing adopted yet — safe no-op
	}
	var supersedes sql.NullInt64
	var origin string
	if err := tx.QueryRow(`SELECT supersedes, origin FROM observations WHERE id=?`, current.Int64).Scan(&supersedes, &origin); err != nil {
		tx.Rollback() //nolint:errcheck
		return fmt.Errorf("resolve abandon doc %d: read supersedes of observation %d: %w", docID, current.Int64, err)
	}
	// Abandon unwinds a RESOLUTION and nothing else. If the merge-entry
	// ResolveAdopt failed (surfaced but the resolver was still entered), the
	// baseline still points at the last genuine 'save'/'load' agreement —
	// deleting THAT row would destroy real observation history and regress
	// the CAS baseline to an older supersedes (review finding: spurious
	// conflicts, or a write blessed against a baseline that reflects no real
	// agreement). Refuse; the caller surfaces the error.
	if origin != "resolve" {
		tx.Rollback() //nolint:errcheck
		return fmt.Errorf("resolve abandon doc %d: baseline observation %d has origin %q, not a resolve adoption — refusing to delete it", docID, current.Int64, origin)
	}
	// Move session_documents.saved_obs OFF the row being deleted FIRST —
	// saved_obs itself carries a FK to observations(id), so deleting the row
	// it still points at would violate that FK before the second statement
	// ever runs.
	if _, err := tx.Exec(`UPDATE session_documents SET saved_obs=? WHERE session_id=? AND doc_id=?`, supersedes, s.sessionID, docID); err != nil {
		tx.Rollback() //nolint:errcheck
		return fmt.Errorf("resolve abandon doc %d: restore saved_obs: %w", docID, err)
	}
	if _, err := tx.Exec(`DELETE FROM observations WHERE id=?`, current.Int64); err != nil {
		tx.Rollback() //nolint:errcheck
		return fmt.Errorf("resolve abandon doc %d: delete observation %d: %w", docID, current.Int64, err)
	}
	if err := tx.Commit(); err != nil {
		return fmt.Errorf("resolve abandon doc %d: commit: %w", docID, err)
	}
	return nil
}

// recordAdoptionTx is the shared one-tx BODY behind every path that moves
// saved_obs to a NEWLY-inserted observation with the given content/stat
// facts: a fresh row (origin, seq, blobHash, and the stat fields), tagged
// with THIS Store's own session_id (v10), is inserted, supersedes is set to
// whatever THIS SESSION's saved_obs held immediately before (NULL if none),
// and this session's session_documents.saved_obs advances to the new row.
// Runs entirely inside the CALLER's already-open tx — recordAdoption below
// wraps this in its own Begin/Commit for standalone callers; commitSave
// (materialize.go) calls it INLINE within its own single save tx, since
// observation + saved_obs + re-Bind must commit atomically together there
// and so cannot be split into a second, independent transaction. A *Store
// method (not a package-level function) so it can read s.sessionID — R3:
// the read (prior saved_obs) and the write (the UPSERT below) both run on
// the SAME tx the caller passed in, preserving the transactional discipline
// CONSTITUTION.md §12 names as what makes flock-free concurrency safe.
func (s *Store) recordAdoptionTx(tx *sql.Tx, docID int64, blobHash string, size int64, mtime string, inode, device, nlink sql.NullInt64, origin string, seq sql.NullInt64, at string) (Observation, error) {
	var supersedes sql.NullInt64
	err := tx.QueryRow(`SELECT saved_obs FROM session_documents WHERE session_id=? AND doc_id=?`, s.sessionID, docID).Scan(&supersedes)
	if err != nil && err != sql.ErrNoRows {
		return Observation{}, fmt.Errorf("read prior saved_obs: %w", err)
	}
	// err == sql.ErrNoRows: no session_documents row yet for THIS session and
	// docID — supersedes stays the zero value (Valid=false), exactly like a
	// never-before-adopted document's saved_obs being NULL under the old
	// documents-column design. The UPSERT below creates the row.
	res, err := tx.Exec(
		`INSERT INTO observations(doc_id, session_id, blob_hash, seq, size, mtime, inode, device, nlink, origin, supersedes, at)
		 VALUES(?,?,?,?,?,?,?,?,?,?,?,?)`,
		docID, s.sessionID, blobHash, seq, size, mtime, inode, device, nlink, origin, supersedes, at,
	)
	if err != nil {
		return Observation{}, fmt.Errorf("insert: %w", err)
	}
	newID, err := res.LastInsertId()
	if err != nil {
		return Observation{}, fmt.Errorf("last insert id: %w", err)
	}
	if _, err := tx.Exec(
		`INSERT INTO session_documents(session_id, doc_id, saved_obs) VALUES(?,?,?)
		 ON CONFLICT(session_id, doc_id) DO UPDATE SET saved_obs=excluded.saved_obs`,
		s.sessionID, docID, newID,
	); err != nil {
		return Observation{}, fmt.Errorf("update saved_obs: %w", err)
	}
	return Observation{
		ID: ObsID(newID), DocID: docID, SessionID: s.sessionID, BlobHash: blobHash, Seq: seq,
		Size: size, MTime: mtime, Inode: inode, Device: device, NLink: nlink,
		Origin: origin, Supersedes: supersedes, At: at,
	}, nil
}

// recordAdoption is the shared one-tx primitive behind every STANDALONE path
// that moves saved_obs to a newly-inserted observation — a reader never
// observes the new observation recorded without saved_obs already pointing
// at it, or vice versa. Shared by AdoptEqual (Probe's crash-recovery
// auto-adopt, F6), ResolveAdopt (a user-driven merge/discard resolution),
// and Load's own first-sighting/heal-adopt cases (F2 item 3). commitSave
// (materialize.go) is NOT a standalone caller — its observation/saved_obs
// move must commit in the SAME tx as its re-Bind, so it calls
// recordAdoptionTx directly inside its own already-open tx instead.
func (s *Store) recordAdoption(docID int64, blobHash string, size int64, mtime string, inode, device, nlink sql.NullInt64, origin string, seq int64) (Observation, error) {
	tx, err := s.perm.Begin()
	if err != nil {
		return Observation{}, fmt.Errorf("record adoption doc %d: begin: %w", docID, err)
	}
	at := s.clock().UTC().Format(time.RFC3339Nano)
	seqVal := sql.NullInt64{Int64: seq, Valid: true}
	obs, err := s.recordAdoptionTx(tx, docID, blobHash, size, mtime, inode, device, nlink, origin, seqVal, at)
	if err != nil {
		tx.Rollback() //nolint:errcheck
		return Observation{}, fmt.Errorf("record adoption doc %d: %w", docID, err)
	}
	if err := tx.Commit(); err != nil {
		return Observation{}, fmt.Errorf("record adoption doc %d: commit: %w", docID, err)
	}
	return obs, nil
}

// AdoptEqual promotes a bare sighting (obs, already recorded — e.g. Probe's
// bare 'probe' observation) to a genuine adoption when its content
// hash-equals the journal-head reconstruction: a NEW origin='resolve'
// observation is inserted, correlated to headSeq (making it
// ANCESTOR-ELIGIBLE, unlike the bare sighting it promotes — F6), and
// saved_obs advances to it. This is the crash-between-swap-and-ack recovery
// path (Materialize's step 5 ack never committed, but the physical swap
// already wrote what the journal head says) — never used for an ordinary
// divergence, which must surface as DiskAhead/Diverged instead.
func (s *Store) AdoptEqual(docID int64, obs ObsID, headSeq int64) (Observation, error) {
	source, err := s.getObservation(obs)
	if err != nil {
		return Observation{}, fmt.Errorf("adopt equal doc %d: read source observation: %w", docID, err)
	}
	adopted, err := s.recordAdoption(docID, source.BlobHash, source.Size, source.MTime, source.Inode, source.Device, source.NLink, "resolve", headSeq)
	if err != nil {
		return Observation{}, fmt.Errorf("adopt equal doc %d: %w", docID, err)
	}
	return adopted, nil
}

File: rust/crates/rune-db/src/adopt.rs
//! The Adoption Contract — the only four verbs allowed to move
//! `session_documents.saved_obs` (`observation.go`'s package doc comment,
//! `observation.go:26-39`): `materialize::commit_save` (inlined, since its
//! move must commit in the SAME tx as its re-Bind), [`adopt_equal`],
//! [`resolve_adopt`], and [`resolve_abandon`]. Port of `pkg/docstate/adopt.go`.

use std::time::SystemTime;

use rusqlite::{Connection, OptionalExtension, Transaction, params};

use crate::Error;
use crate::observation::{self, ObsId, Observation};
use crate::retry;

/// The shared one-tx BODY behind every path that moves `saved_obs` to a
/// NEWLY-inserted observation: a fresh row is inserted (tagged
/// `session_id`), `supersedes` is set to whatever `session_id`'s
/// `saved_obs` held immediately before (`None` if none), and
/// `session_documents.saved_obs` advances to the new row. Runs entirely
/// inside the CALLER's already-open tx — `materialize::commit_save` calls
/// this directly inside its own save tx (observation + saved_obs + re-Bind
/// must commit atomically together there); [`record_adoption`] below wraps
/// it in its own transaction for standalone callers. Port of
/// `adopt.go:101-150` (`recordAdoptionTx`).
#[allow(clippy::too_many_arguments)]
pub(crate) fn record_adoption_tx(
    tx: &Transaction<'_>,
    doc_id: i64,
    session_id: i64,
    blob_hash: &str,
    size: i64,
    mtime: &str,
    inode: Option<i64>,
    device: Option<i64>,
    nlink: Option<i64>,
    origin: &str,
    seq: Option<i64>,
    at: &str,
) -> Result<Observation, Error> {
    let supersedes: Option<i64> = tx
        .query_row(
            "SELECT saved_obs FROM session_documents WHERE session_id=?1 AND doc_id=?2",
            params![session_id, doc_id],
            |r| r.get(0),
        )
        .optional()?
        .flatten();

    tx.execute(
        "INSERT INTO observations(doc_id, session_id, blob_hash, seq, size, mtime, inode, device, nlink, origin, supersedes, at) \
         VALUES(?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12)",
        params![doc_id, session_id, blob_hash, seq, size, mtime, inode, device, nlink, origin, supersedes, at],
    )?;
    let new_id = tx.last_insert_rowid();

    tx.execute(
        "INSERT INTO session_documents(session_id, doc_id, saved_obs) VALUES(?1,?2,?3) \
         ON CONFLICT(session_id, doc_id) DO UPDATE SET saved_obs=excluded.saved_obs",
        params![session_id, doc_id, new_id],
    )?;

    Ok(Observation {
        id: new_id,
        doc_id,
        session_id,
        blob_hash: blob_hash.to_string(),
        seq,
        size,
        mtime: mtime.to_string(),
        inode,
        device,
        nlink,
        origin: origin.to_string(),
        supersedes,
        at: at.to_string(),
    })
}

/// The shared one-tx primitive behind every STANDALONE path that moves
/// `saved_obs` to a newly-inserted observation — [`adopt_equal`],
/// [`resolve_adopt`], and `load::load`'s own first-sighting/heal-adopt
/// cases. Port of `adopt.go:152-177` (`recordAdoption`).
#[allow(clippy::too_many_arguments)]
pub(crate) fn record_adoption(
    conn: &mut Connection,
    doc_id: i64,
    session_id: i64,
    blob_hash: &str,
    size: i64,
    mtime: &str,
    inode: Option<i64>,
    device: Option<i64>,
    nlink: Option<i64>,
    origin: &str,
    seq: i64,
    now: SystemTime,
) -> Result<Observation, Error> {
    let at = crate::session::format_rfc3339_nanos(now);
    let blob_hash = blob_hash.to_string();
    let mtime = mtime.to_string();
    let origin = origin.to_string();
    retry::with_retry(conn, |tx| {
        record_adoption_tx(
            tx,
            doc_id,
            session_id,
            &blob_hash,
            size,
            &mtime,
            inode,
            device,
            nlink,
            &origin,
            Some(seq),
            &at,
        )
    })
}

/// Promotes a bare sighting (`obs`, already recorded — e.g. `probe::probe`'s
/// bare `'probe'` observation) to a genuine adoption when its content
/// hash-equals the journal-head reconstruction: a NEW `origin='resolve'`
/// observation is inserted, correlated to `head_seq` (making it
/// ancestor-eligible, unlike the bare sighting it promotes), and
/// `saved_obs` advances to it. The crash-between-swap-and-ack recovery
/// path — never used for an ordinary divergence. Port of `adopt.go:179-198`
/// (`AdoptEqual`).
pub fn adopt_equal(
    conn: &mut Connection,
    session_id: i64,
    doc_id: i64,
    obs: ObsId,
    head_seq: i64,
    now: SystemTime,
) -> Result<Observation, Error> {
    let source = retry::with_retry(conn, |tx| observation::get_observation(tx, obs))?;
    record_adoption(
        conn,
        doc_id,
        session_id,
        &source.blob_hash,
        source.size,
        &source.mtime,
        source.inode,
        source.device,
        source.nlink,
        "resolve",
        head_seq,
        now,
    )
}

/// Commits a [D]iscard/[M]erge resolution (or an explicit hash-equality
/// adopt): re-tags `obs` as `origin='resolve'`, correlated to `edit_seq`
/// (the seq of the journaled replace-all/merge-entry edit that resolved
/// it), and advances `saved_obs` to it. Undo past `edit_seq` moves the
/// journal position below this resolve observation, so `ancestor_at`
/// automatically stops finding it and `sync` reports `Diverged` again — the
/// guard re-raises with no bespoke unwind logic. Port of `adopt.go:9-31`
/// (`ResolveAdopt`).
pub fn resolve_adopt(
    conn: &mut Connection,
    session_id: i64,
    doc_id: i64,
    obs: ObsId,
    edit_seq: i64,
    now: SystemTime,
) -> Result<Observation, Error> {
    let source = retry::with_retry(conn, |tx| observation::get_observation(tx, obs))?;
    record_adoption(
        conn,
        doc_id,
        session_id,
        &source.blob_hash,
        source.size,
        &source.mtime,
        source.inode,
        source.device,
        source.nlink,
        "resolve",
        edit_seq,
        now,
    )
}

/// Reverses the resolve observation a merge/discard adoption
/// (`resolve_adopt`) or a heal-adopt created — the Esc-abort-out-of-the-
/// merge-resolver counterpart. Reads `doc_id`'s CURRENT `saved_obs`
/// (expected to be the resolve/adoption observation being unwound), deletes
/// that row (the blob is kept — history is never destroyed, only the fact
/// that it was "agreed" is retracted), and restores `saved_obs` to EXACTLY
/// what it superseded. Refuses (surfaced error) if the current `saved_obs`
/// is not itself an `origin='resolve'` row — abandon unwinds a RESOLUTION
/// and nothing else; deleting a genuine `'save'`/`'load'` baseline would
/// destroy real observation history. A doc with no `saved_obs` at all is a
/// safe no-op. Port of `adopt.go:33-99` (`ResolveAbandon`).
pub fn resolve_abandon(conn: &mut Connection, session_id: i64, doc_id: i64) -> Result<(), Error> {
    retry::with_retry(conn, |tx| {
        let current: Option<i64> = tx
            .query_row(
                "SELECT saved_obs FROM session_documents WHERE session_id=?1 AND doc_id=?2",
                params![session_id, doc_id],
                |r| r.get(0),
            )
            .optional()?
            .flatten();
        let Some(current) = current else {
            return Ok(()); // no session_documents row, or saved_obs NULL — nothing adopted yet
        };

        let (supersedes, origin): (Option<i64>, String) = tx.query_row(
            "SELECT supersedes, origin FROM observations WHERE id=?1",
            params![current],
            |r| Ok((r.get(0)?, r.get(1)?)),
        )?;

        if origin != "resolve" {
            return Err(Error::Invalid(format!(
                "resolve abandon doc {doc_id}: baseline observation {current} has origin {origin:?}, not a resolve adoption — refusing to delete it"
            )));
        }

        // Move saved_obs OFF the row being deleted FIRST — it carries a FK
        // to observations(id), so deleting first would violate that FK.
        tx.execute(
            "UPDATE session_documents SET saved_obs=?1 WHERE session_id=?2 AND doc_id=?3",
            params![supersedes, session_id, doc_id],
        )?;
        tx.execute("DELETE FROM observations WHERE id=?1", params![current])?;
        Ok(())
    })
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used, clippy::indexing_slicing)]
mod tests {
    use super::*;

    fn open() -> Connection {
        let conn = Connection::open_in_memory().expect("open");
        crate::schema::apply(&conn).expect("schema");
        conn
    }

    fn seed_doc(conn: &Connection) -> i64 {
        conn.execute(
            "INSERT INTO documents(path, created_at, last_seen_at) VALUES ('', 'x', 'x')",
            [],
        )
        .expect("seed doc");
        conn.last_insert_rowid()
    }

    /// `observations.blob_hash` is FK-constrained to `blobs.hash` — every
    /// hand-seeded observation in these tests needs a real blob row first.
    fn seed_blob(conn: &Connection, content: &str) -> String {
        crate::blob::put_blob(conn, content).expect("seed blob")
    }

    #[test]
    fn resolve_abandon_restores_exact_prior_baseline() {
        let mut conn = open();
        let session_id =
            crate::session::establish_session(&conn, SystemTime::now()).expect("session");
        let doc_id = seed_doc(&conn);
        let hash_1 = seed_blob(&conn, "content 1");

        // First: an ordinary save baseline.
        let first = record_adoption(
            &mut conn,
            doc_id,
            session_id,
            &hash_1,
            1,
            "t",
            None,
            None,
            None,
            "save",
            1,
            SystemTime::now(),
        )
        .expect("first adoption");

        // Then: a resolve adoption on top of it.
        let resolved = adopt_equal(
            &mut conn,
            session_id,
            doc_id,
            first.id,
            2,
            SystemTime::now(),
        )
        .expect("adopt_equal");
        assert_eq!(resolved.supersedes, Some(first.id));

        resolve_abandon(&mut conn, session_id, doc_id).expect("resolve_abandon");

        let current: Option<i64> = conn
            .query_row(
                "SELECT saved_obs FROM session_documents WHERE session_id=?1 AND doc_id=?2",
                params![session_id, doc_id],
                |r| r.get(0),
            )
            .expect("read saved_obs");
        assert_eq!(
            current,
            Some(first.id),
            "must restore the exact prior baseline"
        );

        let deleted: i64 = conn
            .query_row(
                "SELECT COUNT(*) FROM observations WHERE id=?1",
                params![resolved.id],
                |r| r.get(0),
            )
            .expect("count");
        assert_eq!(deleted, 0, "the resolve observation row must be gone");
    }

    #[test]
    fn resolve_abandon_refuses_a_non_resolve_baseline() {
        let mut conn = open();
        let session_id =
            crate::session::establish_session(&conn, SystemTime::now()).expect("session");
        let doc_id = seed_doc(&conn);
        let hash_1 = seed_blob(&conn, "content 1");

        record_adoption(
            &mut conn,
            doc_id,
            session_id,
            &hash_1,
            1,
            "t",
            None,
            None,
            None,
            "save",
            1,
            SystemTime::now(),
        )
        .expect("save adoption");

        let err = resolve_abandon(&mut conn, session_id, doc_id).expect_err("must refuse");
        assert!(matches!(err, Error::Invalid(_)));
    }

    #[test]
    fn resolve_abandon_on_a_document_with_no_baseline_is_a_safe_no_op() {
        let mut conn = open();
        let session_id =
            crate::session::establish_session(&conn, SystemTime::now()).expect("session");
        let doc_id = seed_doc(&conn);

        resolve_abandon(&mut conn, session_id, doc_id).expect("must be a no-op, not an error");
    }
}
